pyincore/analyses/buildingdamage/buildingdamage.py

```diff
@@ -121,66 +122,88 @@ def building_damage_analysis_bulk_input(self, buildings, retrofit_strategy, haza
-                location = GeoUtil.get_location(b)
-                loc = str(location.y) + "," + str(location.x)
-                demands = AnalysisUtil.get_hazard_demand_types(b, fragility_sets[bldg_id], hazard_type)
-                units = fragility_sets[bldg_id].demand_units
-                value = {
-                    "demands": demands,
-                    "units": units,
-                    "loc": loc
-                }
-                values_payload.append(value)
-                mapped_buildings.append(b)
-
-                if use_liquefaction and geology_dataset_id is not None:
-                    value_liq = {
-                        "demands": [""],
-                        "units": [""],
+        multihazard_vals = {}
+        adjust_demand_types_mapping = {}
+
+        for hazard_type, hazard_dataset_id in zip(hazard_types, hazard_dataset_ids):
+            # get allowed demand types for the hazard type
+            allowed_demand_types = [item["demand_type"].lower() for item in self.hazardsvc.get_allowed_demands(
+                hazard_type)]
+
+            # Liquefaction
+            if hazard_type == "earthquake" and self.get_parameter("use_liquefaction") is not None:
+                use_liquefaction = self.get_parameter("use_liquefaction")
+
+            values_payload = []
+            values_payload_liq = []  # for liquefaction, if used
+            unmapped_buildings = []
+            mapped_buildings = []
+            for b in buildings:
+                bldg_id = b["id"]
+                if bldg_id in fragility_sets:
+                    location = GeoUtil.get_location(b)
+                    loc = str(location.y) + "," + str(location.x)
+                    demands, units, adjusted_to_original = \
+                        AnalysisUtil.get_hazard_demand_types_units(b,
+                                                                   fragility_sets[bldg_id],
+                                                                   hazard_type,
+                                                                   allowed_demand_types)
+                    adjust_demand_types_mapping.update(adjusted_to_original)
+                    value = {
+                        "demands": demands,
+                        "units": units,
                         "loc": loc
                     }
-                    values_payload_liq.append(value_liq)
+                    values_payload.append(value)
+                    mapped_buildings.append(b)
+
+                    if use_liquefaction and geology_dataset_id is not None:
+                        value_liq = {
+                            "demands": [""],
+                            "units": [""],
+                            "loc": loc
+                        }
+                        values_payload_liq.append(value_liq)
+                else:
+                    unmapped_buildings.append(b)
+
+            if hazard_type == 'earthquake':
```

Comment:
I thought you were going to add another check here for earthquake+tsunami, but what you did by splitting the types and iterating over the types is probably more generic and better.